torchtune/modules/vq_embeddings.py

```diff
@@ -0,0 +1,173 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Tuple
+
+import torch
+from torch import nn, Tensor
+from torch.nn import functional as F
+
+
+class VectorQuantizedEmbeddings(nn.Module):
+    """
+    Vector quantized embedding layer that takes in the output of an encoder
+    and performs a nearest-neighbor lookup in the embedding space.
+    Vector quantization was introduced in Oord et al. 2017 (https://arxiv.org/pdf/1711.00937.pdf)
+    to generate high-fidelity images, videos, and audio data.
+    The embedding weights are trained with exponential moving average updates as described
+    in original paper.
+
+    Code was adapted from torchmultimodal's `Codebook module
```

Comment:
I was going to comment that I'm more familiar with this being called a codebook. Any reason for not calling it that here? Or VQCodebook?

Reply:
no strong opinions, I've seen it called both. we could do `VectorQuantizedCodebook`?

Reply:
I guess that could be confusing too since it's just a "normal" codebook.  Maybe AutoEncoderCodebook? Or just leave it.

Reply:
Having `VectorQuantized` is the most descriptive imo, so I will leave it as it